src/tools/perf/perftest_daemon.c

```diff
@@ -0,0 +1,243 @@
+#include <unistd.h>
+#include <signal.h>
+
+typedef struct ucp_perf_daemon_context_t {
+    ucp_context_h                      context;
+    ucp_worker_h                       worker;
+    ucp_listener_h                     listener;
+    uint16_t                           port;
+    ucp_ep_h                           client_ep;
+    ucp_ep_h                           peer_ep;
+    struct sockaddr_storage            peer_address;
+    ucp_mem_h                          send_memh;
+    ucp_mem_h                          recv_memh;
+    void                               *rx_address;
+} ucp_perf_daemon_context_t;
+
+static volatile int terminated = 0;
+
+static void
+ucp_perf_daemon_ep_close(ucp_perf_daemon_context_t *ctx, ucp_ep_h ep)
+{
+    ucp_request_param_t param;
+    ucs_status_ptr_t close_req;
+    ucs_status_t status;
+
+    if (NULL == ep) {
+        return;
+    }
+
+    param.op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS;
+    param.flags        = UCP_EP_CLOSE_FLAG_FORCE;
+    close_req          = ucp_ep_close_nbx(ep, &param);
+
+    if (UCS_PTR_IS_PTR(close_req)) {
+        do {
+            ucp_worker_progress(ctx->worker);
+            status = ucp_request_check_status(close_req);
+        } while (status == UCS_INPROGRESS);
+        ucp_request_free(close_req);
+    } else {
+        status = UCS_PTR_STATUS(close_req);
+    }
+
+    if (status != UCS_OK) {
+        ucs_error("daemon failed to close ep %p", ep);
+    }
+}
+
+static void ucp_perf_daemon_err_cb(void *arg, ucp_ep_h ep, ucs_status_t status)
+{
+    terminated = 1;
+}
+
+static void
+ucp_perf_daemon_server_conn_handle_cb(ucp_conn_request_h conn_request,
+                                      void *arg)
+{
+    ucp_perf_daemon_context_t *ctx = arg;
+    ucp_ep_params_t ep_params;
+    ucs_status_t status;
+    ucp_ep_h ep;
+
+    ep_params.field_mask      = UCP_EP_PARAM_FIELD_ERR_HANDLER  |
+                                UCP_EP_PARAM_FIELD_CONN_REQUEST |
+                                UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode        = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.conn_request    = conn_request;
+    ep_params.err_handler.cb  = ucp_perf_daemon_err_cb;
+    ep_params.err_handler.arg = ctx;
+
+    status = ucp_ep_create(ctx->worker, &ep_params, &ep);
+    if (status != UCS_OK) {
+        ucs_error("failed to create an endpoint on the daemon: %s",
+                  ucs_status_string(status));
+    }
+}
+
+static void ucp_perf_daemon_cleanup(ucp_perf_daemon_context_t *ctx)
+{
+    /* coverity[check_return] */
+    ucp_mem_unmap(ctx->context, ctx->send_memh);
+    /* coverity[check_return] */
+    ucp_mem_unmap(ctx->context, ctx->recv_memh);
+
+    ucp_perf_daemon_ep_close(ctx, ctx->peer_ep);
+    ucp_perf_daemon_ep_close(ctx, ctx->client_ep);
+    ucp_listener_destroy(ctx->listener);
+    ucp_worker_destroy(ctx->worker);
+    ucp_cleanup(ctx->context);
+}
+
+static int ucp_perf_daemon_init(ucp_perf_daemon_context_t *ctx)
```

Comment:
better to make it return `ucs_status_t` to make it consistent with `ucp_perf_daemon_parse_cmd`